cdap-watchdog/src/main/java/co/cask/cdap/logging/context/LoggingContextHelper.java

```diff
@@ -189,21 +181,60 @@ public static Filter createFilter(LoggingContext loggingContext) {
+      filterBuilder.add(new MdcExpression(ApplicationLoggingContext.TAG_APPLICATION_ID, applId));
+      filterBuilder.add(new MdcExpression(entityTag.getName(), entityTag.getValue()));
+
+      // Add runid filter if required
+      LoggingContext.SystemTag runId = loggingContext.getSystemTagsMap().get(ApplicationLoggingContext.TAG_RUNID_ID);
+      if (runId != null && runId.getValue() != null) {
+        filterBuilder.add(new MdcExpression(ApplicationLoggingContext.TAG_RUNID_ID, runId.getValue()));
+      }
+
+      // Add adapter filter if required
       if (loggingContext.getSystemTagsMap().containsKey(ApplicationLoggingContext.TAG_ADAPTER_ID)) {
         String adapterName = loggingContext.getSystemTagsMap().get(ApplicationLoggingContext.TAG_ADAPTER_ID).getValue();
-        return new AndFilter(ImmutableList.of(namespaceFilter,
-                                              new MdcExpression(ApplicationLoggingContext.TAG_APPLICATION_ID, applId),
-                                              new MdcExpression(tagName, entityId),
-                                              new MdcExpression(ApplicationLoggingContext.TAG_ADAPTER_ID, adapterName))
-        );
+        filterBuilder.add(new MdcExpression(ApplicationLoggingContext.TAG_ADAPTER_ID, adapterName));
       }
-      return new AndFilter(
-        ImmutableList.of(namespaceFilter,
-                         new MdcExpression(ApplicationLoggingContext.TAG_APPLICATION_ID, applId),
-                         new MdcExpression(tagName, entityId)
-        )
-      );
+      return new AndFilter(filterBuilder.build());
+    }
+  }
 
+  public static LoggingContext.SystemTag getEntityId(LoggingContext loggingContext) {
+    final String tagName;
+    final String entityId;
+    if (loggingContext instanceof FlowletLoggingContext) {
+      tagName = FlowletLoggingContext.TAG_FLOW_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
+    } else if (loggingContext instanceof ProcedureLoggingContext) {
+      tagName = ProcedureLoggingContext.TAG_PROCEDURE_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
+    } else if (loggingContext instanceof MapReduceLoggingContext) {
+      tagName = MapReduceLoggingContext.TAG_MAP_REDUCE_JOB_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
+    } else if (loggingContext instanceof SparkLoggingContext) {
+      tagName = SparkLoggingContext.TAG_SPARK_JOB_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
+    } else if (loggingContext instanceof UserServiceLoggingContext) {
+      tagName = UserServiceLoggingContext.TAG_USERSERVICE_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
+    } else if (loggingContext instanceof WorkerLoggingContext) {
+      tagName = WorkerLoggingContext.TAG_WORKER_ID;
+      entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();
```

Comment:
``entityId = loggingContext.getSystemTagsMap().get(tagName).getValue();`` is common in all conditions, may be we can move that outside.